onnx/defs/tensor/defs.cc

```diff
@@ -77,19 +77,126 @@ could also be 0, in which case the actual dimension value is unchanged (i.e. tak
+          const auto &dataInputTensorType = ctx.getInputType(0)->tensor_type();
+          std::vector<bool> unresolvedZeros(targetShape.size(), false);
+          int64_t outputProduct = 1;
+          for (int i = 0; i < static_cast<int>(targetShape.size()); ++i) {
+            // Add a new dimension to outputShape
+            auto *new_dim = outputShape->add_dim();
+            if (targetShape[i] == -1) {
+              // Check if multiple -1's. If not, set negativeOneDim, marking
+              // this dimension to potentially be filled in later.
+              if (negativeOneDim) {
+                fail_shape_inference(
+                    "Target shape may not have multiple -1 dimensions");
+              }
+              negativeOneDim = new_dim;
+            } else if (targetShape[i] == 0) {
+              // Check if data input has a shape and if the index i is within
+              // its bounds. If these conditions are satisfied, any dimension
+              // value/param should be propogated. If dimension value cannot be
+              // inferred, set the corresponding  unresolvedZeros flag to true.
+              unresolvedZeros[i] = true;
+              if (dataInputTensorType.has_shape() &&
+                  i < dataInputTensorType.shape().dim_size()) {
+                if (dataInputTensorType.shape().dim(i).has_dim_value()) {
+                  const auto &dim_value =
+                      dataInputTensorType.shape().dim(i).dim_value();
+                  new_dim->set_dim_value(dim_value);
+                  outputProduct *= dim_value;
+                  unresolvedZeros[i] = false;
+                } else if (dataInputTensorType.shape().dim(i).has_dim_param()) {
+                  const auto &dim_param =
+                      dataInputTensorType.shape().dim(i).dim_param();
+                  new_dim->set_dim_param(dim_param);
+                }
+              }
+            } else if (targetShape[i] > 0) {
+              // Set the dimension value to targetShape[i]
+              new_dim->set_dim_value(targetShape[i]);
+              outputProduct *= targetShape[i];
+            } else {
+              // Check if value is less than -1; fail if so
+              fail_shape_inference("Dimension value is invalid");
+            }
+          }
+
+          // If negativeOneDim has been set, we attempt to infer its value. This
+          // can be done if all dimension values for the data input tensor shape
+          // are known other than the ones corresponding to unresolvedZeros
+          // flags.
+          if (negativeOneDim) {
+            // First, attempt to compute product of data input shape dimensions
+            // that are not marked by unresolvedZeros. If not possible, set the
+            // inputProductValid flag to false.
+            int64_t inputProduct = 1;
+            bool inputProductValid = true;
+            if (!dataInputTensorType.has_shape()) {
```

Comment:
add "i >= static_cast<int>(unresolvedZeros.size()" here.

Reply:
@linkerzhang what do you mean by this? We're not in a loop here?